File: python/item.py
import pandas as pd
import json


def df_to_json(df):

    output = {}

    for _, row in df.iterrows():
        obj = {}
        for key, val in row.items():
            if key.startswith("Unnamed:") or pd.isna(val) or val == "":
                continue

            if key == 'id':
                id = val
            elif key == 'gold':
                obj[key] = int(val)
            elif key == 'icon' or key == 'cat':
                obj[key] = val
            elif key.startswith("_"):
                obj.update(json.loads(f"{{{val}}}"))
            elif key == 'effects' or key == 'procs':
                obj[key] = json.loads(f"[{val}]")
            elif key == 'bb' or key.startswith('bb_'):
                obj.setdefault('bb', {}).update(json.loads(f"{{{val}}}"))
            else:
                obj[key] = json.loads(f"{{{val}}}")
        if obj:
            output[id] = obj

    return output


def excel_to_json(input_excel_path, output_json_path, all_sheets=True):
    output = {}
    if all_sheets:
        excel_data = pd.read_excel(input_excel_path, sheet_name=None)
        for _, df in excel_data.items():
            output.update(df_to_json(df))
    else:
        df = pd.read_excel(input_excel_path)
        output = df_to_json(df)

    with open(output_json_path, "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print("item 轉換完成 ✅")


def unit_test():
    input_excel_path = "./xls/item.xlsx"
    output_json_path = "./public/assets/json/item.json"
    excel_to_json(input_excel_path, output_json_path)


if __name__ == "__main__":
    unit_test()

File: scripts/item.js
import XLSX from 'xlsx';
import { writeFileSync } from 'fs';

const STR_COLS = new Set(['icon', 'cat']);
const ARR_COLS = new Set(['effects', 'procs']);

function sheetToJson(ws) {
  const allRows = XLSX.utils.sheet_to_json(ws, { header: 1, defval: null });
  if (!allRows.length) return {};
  const headers = allRows[0].map(h => (h !== null && h !== undefined ? String(h) : null));
  const output = {};

  for (let r = 1; r < allRows.length; r++) {
    const row = allRows[r];
    const obj = {};
    let id;

    for (let i = 0; i < headers.length; i++) {
      const key = headers[i];
      if (!key || key.startsWith('Unnamed:')) continue;
      const val = row[i];
      if (val === null || val === undefined || val === '') continue;

      const strVal = String(val);

      if (key === 'id') {
        id = strVal;
      } else if (key === 'gold') {
        obj[key] = Math.round(Number(val));
      } else if (STR_COLS.has(key)) {
        obj[key] = strVal;
      } else if (key.startsWith('_')) {
        Object.assign(obj, JSON.parse('{' + strVal + '}'));
      } else if (ARR_COLS.has(key)) {
        obj[key] = JSON.parse('[' + strVal + ']');
      } else if (key === 'bb' || key.startsWith('bb_')) {
        if (!obj.bb) obj.bb = {};
        Object.assign(obj.bb, JSON.parse('{' + strVal + '}'));
      } else {
        obj[key] = JSON.parse('{' + strVal + '}');
      }
    }

    if (id !== undefined && Object.keys(obj).length > 0) output[id] = obj;
  }
  return output;
}

function excelToJson(inputPath, outputPath, allSheets = true) {
  const wb = XLSX.readFile(inputPath);
  const output = {};
  const names = allSheets ? wb.SheetNames : [wb.SheetNames[0]];
  for (const name of names) {
    Object.assign(output, sheetToJson(wb.Sheets[name]));
  }
  writeFileSync(outputPath, JSON.stringify(output, null, 2), 'utf-8');
  console.log('item 轉換完成');
}

excelToJson('./xls/item.xlsx', './public/assets/json/item.json');
